cedar-policy-core/src/est/expr.rs

```diff
@@ -536,59 +542,64 @@ impl TryFrom<Expr> for ast::Expr {
             )),
             Expr::ExprNoExt(ExprNoExt::ContainsAll { left, right }) => Ok(ast::Expr::contains_all(
-                (*left).clone().try_into()?,
-                (*right).clone().try_into()?,
+                (*left).clone().try_into_ast(id.clone())?,
+                (*right).clone().try_into_ast(id)?,
             )),
             Expr::ExprNoExt(ExprNoExt::ContainsAny { left, right }) => Ok(ast::Expr::contains_any(
-                (*left).clone().try_into()?,
-                (*right).clone().try_into()?,
+                (*left).clone().try_into_ast(id.clone())?,
+                (*right).clone().try_into_ast(id)?,
             )),
             Expr::ExprNoExt(ExprNoExt::GetAttr { left, attr }) => {
-                Ok(ast::Expr::get_attr((*left).clone().try_into()?, attr))
+                Ok(ast::Expr::get_attr((*left).clone().try_into_ast(id)?, attr))
             }
             Expr::ExprNoExt(ExprNoExt::HasAttr { left, attr }) => {
-                Ok(ast::Expr::has_attr((*left).clone().try_into()?, attr))
+                Ok(ast::Expr::has_attr((*left).clone().try_into_ast(id)?, attr))
             }
             Expr::ExprNoExt(ExprNoExt::Like { left, pattern }) => {
                 match unescape::to_pattern(&pattern) {
-                    Ok(pattern) => Ok(ast::Expr::like((*left).clone().try_into()?, pattern)),
-                    Err(errs) => Err(Self::Error::UnescapeError(errs)),
+                    Ok(pattern) => Ok(ast::Expr::like((*left).clone().try_into_ast(id)?, pattern)),
+                    Err(errs) => Err(FromJsonError::UnescapeError(errs)),
                 }
             }
             Expr::ExprNoExt(ExprNoExt::If {
                 cond_expr,
                 then_expr,
                 else_expr,
             }) => Ok(ast::Expr::ite(
-                (*cond_expr).clone().try_into()?,
-                (*then_expr).clone().try_into()?,
-                (*else_expr).clone().try_into()?,
+                (*cond_expr).clone().try_into_ast(id.clone())?,
+                (*then_expr).clone().try_into_ast(id.clone())?,
+                (*else_expr).clone().try_into_ast(id)?,
             )),
             Expr::ExprNoExt(ExprNoExt::Set(elements)) => Ok(ast::Expr::set(
                 elements
                     .into_iter()
-                    .map(|el| el.try_into())
-                    .collect::<Result<Vec<_>, Self::Error>>()?,
-            )),
-            Expr::ExprNoExt(ExprNoExt::Record(map)) => Ok(ast::Expr::record(
-                map.into_iter()
-                    .map(|(k, v)| Ok((k, v.try_into()?)))
-                    .collect::<Result<HashMap<SmolStr, _>, Self::Error>>()?,
+                    .map(|el| el.try_into_ast(id.clone()))
+                    .collect::<Result<Vec<_>, FromJsonError>>()?,
             )),
+            Expr::ExprNoExt(ExprNoExt::Record(map)) => {
+                // PANIC SAFETY: can't have duplicate keys here because the input was already a HashMap
```

Comment:
Most of the `expect`s here are asserting that `.iter().map()` (and similar operation on `unzip()` or on `keys()` and `values()`) affecting only values without touching keys cannot introduce duplicates. Maybe this could be abstracted into a utility function.

Reply:
I think adding abstraction is just as likely to obscure potential bugs.  With this design, although we have more `expect`s, each `expect` can be verified with local reasoning about the ~10 lines of code surrounding it